Simplified traits.

gabrik opened the pull request:
The idea behind this PR is to simplify the usage of Zenoh-Flow interface when creating nodes.

**Before:**
```rust



#[derive(Debug)]
struct SumAndSendFactory;

struct SumAndSend {
    input: Arc<Input>,
    output: Arc<Output>,
    counter: Arc<AtomicUsize>,
}

static INPUT: &str = "Number";
static OUTPUT: &str = "Sum";

#[async_trait]
impl OperatorFactoryTrait for SumAndSendFactory {
    async fn new_operator(
        &self,
        _context: &mut Context,
        _configuration: &Option<Configuration>,
        mut inputs: Inputs,
        mut outputs: Outputs,
    ) -> Result<Option<Arc<dyn Node>>> {
        let input = inputs.take_into_arc(INPUT).unwrap();
        let output = outputs.take_into_arc(OUTPUT).unwrap();
        let counter = Arc::new(AtomicUsize::new(0));

        Ok(Some(Arc::new(SumAndSend {
            input,
            output,
            counter,
        })))
    }
}

#[async_trait]
impl Node for SumAndSend {
    async fn iteration(&self) -> Result<()> {
        let data = if let Message::Data(mut message) = self.input.recv_async().await.unwrap() {
            message.get_inner_data().try_get::<ZFUsize>()?.clone()
        } else {
            bail!(ErrorKind::Empty, "No data")
        };

        self.counter.fetch_add(data.0, Ordering::AcqRel);
        let data = Data::from(ZFUsize(self.counter.load(Ordering::Acquire)));

        self.output.send_async(data, None).await
    }
}

zenoh_flow::export_operator_factory!(register);

fn register() -> Result<Arc<dyn OperatorFactoryTrait>> {
    Ok(Arc::new(SumAndSendFactory) as Arc<dyn OperatorFactoryTrait>)
}

```

**After**

```rust
#[export_operator]
struct SumAndSend {
    input: Arc<Input>,
    output: Arc<Output>,
    counter: Arc<AtomicUsize>,
}

static INPUT: &str = "Number";
static OUTPUT: &str = "Sum";

#[async_trait]
impl Operator for SumAndSend {

    fn new(
         _context: &mut Context,
        _configuration: &Option<Configuration>,
        mut inputs: Inputs,
        mut outputs: Outputs,
    )-> Result<Option<Self>> {
            Ok(Some(SumAndSend {
                input,
                output,
                counter,
            }))
    }
}

#[async_trait]
impl Node for SumAndSend {
    async fn iterartion(&self) -> Result<()> {
        let data = if let Message::Data(mut message) = self.input.recv_async().await.unwrap() {
            message.get_inner_data().try_get::<ZFUsize>()?.clone()
        } else {
            bail!(ErrorKind::Empty, "No data")
        };
        
        self.counter.fetch_add(data.0, Ordering::AcqRel);
        let data = Data::from(ZFUsize(self.counter.load(Ordering::Acquire)));

        self.output.send_async(data, None).await
    
    }
}

```

gabrik committed:
WIP: initial re-refacto of node traits

Signed-off-by: gabrik <gabriele.baldoni@gmail.com>

gabrik committed:
WIP: removed old traits, added docs for new traits, updated dataflow test, cleanup

Signed-off-by: gabrik <gabriele.baldoni@gmail.com>

gabrik committed:
update macros

Signed-off-by: gabrik <gabriele.baldoni@gmail.com>

J-Loudet commented on zenoh-flow/src/macros.rs:
```diff
@@ -65,40 +76,47 @@ macro_rules! export_operator_factory {
 
 /// This macros should be used in order to provide the symbols
-/// for the dynamic load of a Sink. Along with a register function
+/// for the dynamic load of an Source. Along with a register function
```
s/an/a

J-Loudet commented on zenoh-flow/src/traits.rs:
```diff
@@ -99,253 +99,310 @@ pub trait ZFData: DowncastAny + Debug + Send + Sync {
+    fn from(other: Arc<dyn Source>) -> Self {
+        Self(other)
+    }
+}
```
Having:
```rust
pub(crate) struct CastSource(pub(crate) Arc<dyn Source>);
```
would remove the need to have an `impl From` and allow us to do:
```rust
let cast_source = CastSource(source);
```

J-Loudet commented on zenoh-flow/src/traits.rs:
```diff
@@ -99,253 +99,310 @@ pub trait ZFData: DowncastAny + Debug + Send + Sync {
+    fn from(other: Arc<dyn Sink>) -> Self {
+        Self(other)
+    }
+}
```
Same comment

J-Loudet commented on zenoh-flow/src/traits.rs:
```diff
@@ -99,253 +99,310 @@ pub trait ZFData: DowncastAny + Debug + Send + Sync {
+    fn from(other: Arc<dyn Operator>) -> Self {
+        Self(other)
+    }
+}
```
Same comment

J-Loudet commented on zenoh-flow/src/traits.rs:
```diff
@@ -99,253 +99,310 @@ pub trait ZFData: DowncastAny + Debug + Send + Sync {
+    async fn iteration(&self) -> ZFResult<()>;
+}
+
+#[async_trait]
+impl Node for dyn Source {
+    async fn iteration(&self) -> ZFResult<()> {
+        Source::iteration(self).await
+    }
+}
+
+#[async_trait]
+impl Node for dyn Sink {
+    async fn iteration(&self) -> ZFResult<()> {
+        Sink::iteration(self).await
+    }
+}
+
+#[async_trait]
+impl Node for dyn Operator {
+    async fn iteration(&self) -> ZFResult<()> {
+        Operator::iteration(self).await
+    }
 }
```
Being perfectly aware that this was my suggestion, I feel the need to ask: is this of any use in the end?

gabrik replied:
If commented it does not compile, so it is quite useful

gabrik committed:
updated loader and macros

Signed-off-by: gabrik <gabriele.baldoni@gmail.com>

J-Loudet reviewed (commented):
Small feedback

gabrik committed:
Addressing comments

Signed-off-by: gabrik <gabriele.baldoni@gmail.com>

gabrik committed:
Fix typo

Signed-off-by: gabrik <gabriele.baldoni@gmail.com>

gabrik committed:
reorganized Source, Sink, Operator and Node traits

Signed-off-by: gabrik <gabriele.baldoni@gmail.com>

gabrik committed:
hidden Factory, new derive macros for export (added to the prelude)

Signed-off-by: gabrik <gabriele.baldoni@gmail.com>

Commit:
refacto: remove Callback shorthand

Supporting the "transformation" of Input / Output into Callback introduced
constraints in our API (and internal codebase). This PR removes this support to
simplify our API.

One of the main benefit is the signature of the "constructors":

before:
```rust
fn new(
   context: &mut Context,
   configuration: &Option<Configuration>,
   mut outputs: Outputs
   ) -> Result<Option<Self>>;
```

after:
```rust
fn new(
   context: &mut Context,
   configuration: &Option<Configuration>,
   mut outputs: Outputs
   ) -> Result<Self>;
```

Commit:
refacto: replace internal Factory by constructors

By internally mapping the `new` static method of each Node to a "constructor"
function (e.g. `SourceFn`) we remove the need of the additional trait `Factory`.

To make each constructor asynchronous while not staining the entire API with
lifetime annotations, the different Node traits have been modified to take no
reference: `&mut Context` became `Context` and `&Option<Configuration>` became
`Option<Configuration>`.

Commit:
fix: typo Deseralization -> Deserialization

gabrik committed:
using attribute macros for exporting source/sink/opererators

Signed-off-by: gabrik <gabriele.baldoni@gmail.com>

gabrik committed:
fix export_operator procedural macro

Signed-off-by: gabrik <gabriele.baldoni@gmail.com>

gabrik committed:
fix doc strings

Signed-off-by: gabrik <gabriele.baldoni@gmail.com>

J-Loudet reviewed (approved):
LGTM!
